### PR TITLE
User dashboard

### LINKED ISSUE #18: User dashboard
- [ ] List of tasks assigned to current user
- [ ] List of user's projects
- [ ] Action in `HomeController`
- [ ] Route for dashboard

### PR DESCRIPTION
**Pull request by neonowy**
* Added UsersController with dashboard action.
* Added dashboard view.
* Added dashboard styles.
* Added json requests handling for `tasks#mark_as_done`.
* Added `Task#form_id` helper for checkbox ids in view.
* Added User#projects for all user's projects.
* Added reqwest js library for ajax.
* Improved tests.
* Tweaked existing styles.

Closes #18.

### CODE REVIEW

**Comment by pospieszynski** [comment]
:+1: 
@gotar

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class UsersController < ApplicationController
+  def dashboard
+    @user_tasks = current_user.assigned_tasks.where(done: false).group_by(&:project)
```
Comment (gotar) [inline app/controllers/users_controller.rb]: will be better to create a scope in a model (query should not leak outside model), it;s a bad practice. Define a scope and use it here

File: app/assets/javascripts/dashboard.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+function onTaskTitleClick(e) {
+  var taskItem   = e.target.parentNode,
+      checkboxId = e.target.id || e.target.getAttribute("for"),
+      projectId  = checkboxId.split("-")[0],
+      taskId     = checkboxId.split("-")[1];
+
+      csrfToken = document.querySelector("meta[name=csrf-token]")
+                  .getAttribute("content");
+
+  reqwest({
```
Comment (smt116) [inline app/assets/javascripts/dashboard.js]: I guess that this `regwest` lib is only for ajax requests, right? Why do you need it at all if you already have jquery in the project? It provides interface for this kind of job

File: app/models/task.rb [hunk 3]
```diff
@@ -7,6 +7,10 @@ class Task < ApplicationRecord
 
   belongs_to :project
 
+  def form_id
+    "#{project.id}-#{id}"
+  end
```
Comment (smt116) [inline app/models/task.rb]: is it really really needed in the model? maybe a view helper would be a better place?

File: app/models/user.rb [hunk 4]
```diff
@@ -16,11 +16,10 @@ class User < ApplicationRecord
...
-      return true
-    elsif self.contributed_projects.any? { |project| project.id == project_id }
-      return true
-    end
-    false
+    projects.any? { |project| project.id == project_id }
+  end
+
+  def projects
+    self.managed_projects + self.contributed_projects
```
Comment (smt116) [inline app/models/user.rb]: `self` is needed here?

File: app/assets/javascripts/dashboard.js (hunk 2 above)
Comment (neonowy) [inline app/assets/javascripts/dashboard.js]: Yes, I know that jquery could do that but I want to get rid of jquery completely in the future.

Comment (smt116) [inline app/assets/javascripts/dashboard.js]: you want to do this in the future so you are adding another lib? :P.. How far this future is? I mean this is something in the backlog that will happen in really near future or this is a plan for the future development and you need to discuss this with the team?

Comment (neonowy) [inline app/assets/javascripts/dashboard.js]: @smt116  It's not yet in the backlog, but it'll happen when I change project tasks view to looks more like a dashboard (jquery is only used in non GET links), so it'll happen in really near future.

Comment (smt116) [inline app/assets/javascripts/dashboard.js]: Okay, I am not familiar with your plans. Just make sure that rest of the team agree on the change. Still not sure what is the reason of removing jquery.

File: app/models/user.rb (hunk 4 above)
Comment (neonowy) [inline app/models/user.rb]: It was left from previous code but it turns out it's not really needed. Thanks :)

**Comment by karolcircle** [comment]
:+1:

**Comment by gotar** [comment]
ok you can merge

**Pull request merged**